sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala

```diff
@@ -0,0 +1,126 @@
+           |Data columns: ${query.output.map(c => s"'${c.name}'").mkString(", ")}""".stripMargin)
+    }
+
+    val errors = new mutable.ArrayBuffer[String]()
+    val resolved: Seq[NamedExpression] = if (byName) {
+      expected.flatMap { tableAttr =>
+        query.resolve(Seq(tableAttr.name), conf.resolver) match {
+          case Some(queryExpr) =>
+            checkField(tableAttr, queryExpr, byName, conf, err => errors += err)
+          case None =>
+            errors += s"Cannot find data for output column '${tableAttr.name}'"
+            None
+        }
+      }
+
+    } else {
+      if (expected.size > query.output.size) {
+        throw new AnalysisException(
+          s"""Cannot write to '$tableName', not enough data columns:
+             |Table columns: ${expected.map(c => s"'${c.name}'").mkString(", ")}
+             |Data columns: ${query.output.map(c => s"'${c.name}'").mkString(", ")}"""
+            .stripMargin)
+      }
+
+      query.output.zip(expected).flatMap {
+        case (queryExpr, tableAttr) =>
+          checkField(tableAttr, queryExpr, byName, conf, err => errors += err)
+      }
+    }
+
+    if (errors.nonEmpty) {
+      throw new AnalysisException(
+        s"Cannot write incompatible data to table '$tableName':\n- ${errors.mkString("\n- ")}")
+    }
+
+    if (resolved == query.output) {
+      query
+    } else {
+      Project(resolved, query)
+    }
+  }
+
+  private def checkField(
+      tableAttr: Attribute,
+      queryExpr: NamedExpression,
+      byName: Boolean,
+      conf: SQLConf,
+      addError: String => Unit): Option[NamedExpression] = {
+
+    lazy val outputField = if (tableAttr.dataType.sameType(queryExpr.dataType) &&
+      tableAttr.name == queryExpr.name &&
+      tableAttr.metadata == queryExpr.metadata) {
+      Some(queryExpr)
+    } else {
+      // Renaming is needed for handling the following cases like
+      // 1) Column names/types do not match, e.g., INSERT INTO TABLE tab1 SELECT 1, 2
+      // 2) Target tables have column metadata
+      Some(Alias(
+        Cast(queryExpr, tableAttr.dataType, Option(conf.sessionLocalTimeZone)),
+        tableAttr.name)(explicitMetadata = Option(tableAttr.metadata)))
+    }
+
+    conf.storeAssignmentPolicy match {
+      case StoreAssignmentPolicy.LEGACY =>
+        outputField
+
+      case StoreAssignmentPolicy.STRICT =>
+        // run the type check first to ensure type errors are present
+        val canWrite = DataType.canWrite(
+          queryExpr.dataType, tableAttr.dataType, byName, conf.resolver, tableAttr.name, addError)
+        if (queryExpr.nullable && !tableAttr.nullable) {
+          addError(s"Cannot write nullable values to non-null column '${tableAttr.name}'")
+          None
+
+        } else if (!canWrite) {
+          None
+
+        } else {
+          outputField
+        }
+
+      case other =>
+        throw new AnalysisException(s"Unsupported store assignment policy: $other")
```

Comment:
We need this? It seems we already have checked if the mode is valid? https://github.com/apache/spark/pull/25453/files#diff-9a6b543db706f1a90f790783d6930a13R1661